src/_purryOrderRules.ts

```diff
@@ -0,0 +1,147 @@
+ * the data, and a compare function that can be used to determine the order
+ * between the items of the array.
+ * This functions takes care of the rest; it will parse rules, built the
+ * comparer, and manage the purrying of the input arguments.
+ */
+export function purryOrderRules<T>(
+  func: (data: ReadonlyArray<T>, compareFn: CompareFunction<T>) => unknown,
+  inputArgs: IArguments | ReadonlyArray<unknown>
+): unknown {
+  // We rely on casting blindly here, but we rely on casting blindly everywhere
+  // else when we call purry so it's fine...
+  const [dataOrRule, ...rules] = (
+    Array.isArray(inputArgs) ? inputArgs : Array.from(inputArgs)
+  ) as
+    | [data: ReadonlyArray<T>, ...rules: Readonly<NonEmptyArray<OrderRule<T>>>]
+    | Readonly<NonEmptyArray<OrderRule<T>>>;
+
+  if (!isOrderRule(dataOrRule)) {
+    // dataFirst!
+
+    // @ts-expect-error [ts2556]: Typescript is failing to infer the type of rules
+    // correctly here after the type refinement above, rules should be non-empty
+    // when we get here.
+    const compareFn = orderRuleComparer(...rules);
+    return func(dataOrRule, compareFn);
+  }
+
+  // dataLast!
+
+  // Important: initialize the comparer outside of the returned function so it
+  // it's constructed and shared everywhere (it's stateless so should be safe
+  // if used multiple times).
+  const compareFn = orderRuleComparer(dataOrRule, ...rules);
+  return (data: ReadonlyArray<T>) => func(data, compareFn);
+}
+
+/**
+ * Some functions need an extra number argument, this helps facilitate that.
+ */
+export function purryOrderRulesWithNumberArgument<T>(
+  func: (
+    data: ReadonlyArray<T>,
+    compareFn: CompareFunction<T>,
+    n: number
+  ) => unknown,
+  inputArgs: IArguments
+): unknown {
+  const [first, second, ...rest] = Array.from(inputArgs);
+
+  // We need to pull the `n` argument out to make it work with purryOrderRules.
+  let n: number;
+  let args: ReadonlyArray<unknown>;
+  if (typeof first === 'number') {
+    // dataLast!
+    n = first;
+    args = [second, ...rest];
+  } else if (typeof second === 'number') {
+    // dataFirst!
+    n = second;
+    args = [first, ...rest];
+  } else {
+    throw new Error("Couldn't find a number argument in the called arguments");
+  }
+
+  return purryOrderRules<T>((...args) => func(...args, n), args);
+}
+
+function orderRuleComparer<T>(
+  primaryRule: OrderRule<T>,
+  secondaryRule?: OrderRule<T>,
+  ...otherRules: ReadonlyArray<OrderRule<T>>
+): (a: T, b: T) => number {
+  const projector =
+    typeof primaryRule === 'function' ? primaryRule : primaryRule[0];
+
+  const direction = typeof primaryRule === 'function' ? 'asc' : primaryRule[1];
+  const comparator = COMPARATOR[direction];
+
+  const nextComparer =
+    secondaryRule === undefined
+      ? undefined
+      : orderRuleComparer(secondaryRule, ...otherRules);
+
+  return (a, b) => {
+    const projectedA = projector(a);
+    const projectedB = projector(b);
```

Comment:
**question:** this means we may invoke `projector` more than once per element, right?

the guarantee that `projector` is invoked at most once per element isn't necessary, and it's not even something most libraries give... the only one i can think of that guarantees this is [deno std](https://deno.land/std@0.208.0/collections/mod.ts?s=sortBy). so not necessary

Reply:
It's not the job of this utility to optimize the projector function. The user can provide a function that also caches its results if they want to. If we did this automatically without the user's control, we could get to a situation where a simple projection (like `identity`) would result in doubling the array's memory footprint; if you are sorting a huge array, that is just redundant memory.